### ISSUE TITLE
Feature request: Make preprocess_weights_for_loading() more flexible

### ISSUE THREAD

**Issue by jlherren** [keras-team/keras#11071]
Currently it's possible to load models built and saved with the efficient cuDNN-based layers (for example `CuDNNLSTM`) into regular non-cuDNN layers (for example `LSTM`). The shapes of the weights in these implementations differ on purpose to be able to do the necessary shuffling at load time within `preprocess_weights_for_loading()`. However it detects only a small set of situations where the RNN is inside one of `Bidirectional`, `TimeDistributed`, `Model` or `Sequence`. This magic does not work in any other situation.

One concrete situation is shown below with a custom `Wrapper` sub-class. However this feature request isn't limited to `Wrapper` sub-classes or `Model` sub-classes. It would ideally be possible to register custom hooks to do weight transformations at load time for any customer layers.

Here's another situation where this would have been useful: I had a custom stateful layer for which I wanted the states to be saved to a file as part of the model. I therefore declared them to be non-trainable weights. However when changing the batch size the shape of the states would change accordingly and it was impossible to load the model back from disk, due to the shape mismatch. What I would like is that those states are loaded only if the shape matches and just discarded otherwise, instead of throwing an error. A custom weight-loading-callback would have been useful in that situation. (Fortunately I found another solution for this situation.)

Here's the full concrete example with the `Wrapper` sub-class:

```python
from keras import Sequential
from keras.layers import CuDNNLSTM, LSTM, Wrapper


class ResidualWrapper(Wrapper):
    """Simple wrapper that outputs the sum of the input and output of the inner layer"""

    def build(self, input_shape=None):
        self.layer.build(input_shape)
        super().build(input_shape)

    def call(self, inputs, **kwargs):
        output = self.layer(inputs, **kwargs)
        return inputs + output

    def compute_output_shape(self, input_shape):
        output_shape = self.layer.compute_output_shape(input_shape)
        if output_shape != input_shape:
            raise ValueError('ResidualWrapper needs child layer to produce output with same shape as input,'
                             'got %r -> %r' % (input_shape, output_shape))
        return output_shape


# This works fine

print('Creating cudnn model for training')
model = Sequential([
    CuDNNLSTM(units=5, return_sequences=True, input_shape=(7, 5))
])
model.save_weights('test.h5')

print('Reloading as non-cudnn model for inference')
model = Sequential([
    LSTM(units=5, return_sequences=True, input_shape=(7, 5))
])
model.load_weights('test.h5')

# This does not

print('Creating cudnn model for training')
model = Sequential([
    ResidualWrapper(CuDNNLSTM(units=5, return_sequences=True), input_shape=(7, 5))
])
model.save_weights('test.h5')

print('Reloading as non-cudnn model for inference')
model = Sequential([
    ResidualWrapper(LSTM(units=5, return_sequences=True), input_shape=(7, 5))
])
model.load_weights('test.h5')
```

Perhaps a `preprocess_weights()` method on `Layer` would be a way to go.
